TDL15-CloudFormation-Jenkins/cf-deploy-jenkins.yaml

```diff
@@ -0,0 +1,245 @@
+Description: AWS CloudFormation Jenkins Template to create a Stand Alone Jenkins server stack using a single EC2 instance with an attached EBS volume that will be used for /var/lib/jenkins
+# Define Parameter Variables that will be used throughout this Cloudformation template.
+Parameters:
+  InstanceType:
+    Description: Type of EC2 instance to launch for the server. 
+    Type: String
+    Default: t2.micro
+    ConstraintDescription: Must be a valid EC2 instance type
+    AllowedValues: 
+      - t2.nano
+      - t2.micro
+      - t2.small
+      - t2.medium
+      - t2.large
+      - t2.xlarge
+      - t2.2xlarge
+      - m4.large
+      - m4.xlarge
+      - m4.2xlarge
+      - m4.4xlarge
+      - m4.10xlarge
+      - m4.16xlarge
+      - m5.large
+      - m5.xlarge
+      - m5.2xlarge
+      - m5.4xlarge
+      - m5.12xlarge
+      - m5.24xlarge
+  VpcId:
+    Description: VPC ID that this stack will be launched in.
+    Type: AWS::EC2::VPC::Id
+    Default: "vpc-02ec2836691642ace"
+    AllowedPattern: "[a-z0-9-]*"
+  SubnetId:
+    Description: VPC Subnet that this stack will be launched in.
+    Type: AWS::EC2::Subnet::Id
+    Default: "subnet-04709dce111382a62"
+    AllowedPattern: "[a-z0-9-]*"
+  KeyName:
+    Description: Name of an existing EC2 KeyPair to enable SSH access to the instance(s).
+    Type: AWS::EC2::KeyPair::KeyName
+    Default: "da-key"
+    ConstraintDescription: Must be the name of an existing EC2 KeyPair
+  SSHLocation:
+    Description: The source IP address (/32) or source IP address range (x.x.x.x/x) that will be allowed to SSH to the EC2 instances
+    Type: String
+    MinLength: 9
+    MaxLength: 18
+    Default: 0.0.0.0/0
+    AllowedPattern: "(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})/(\\d{1,2})"
+    ConstraintDescription: Must be a valid IP CIDR range of the form x.x.x.x/x
+  JenkinsPort:
+    Description: The Port that will be used to access Jenkins (Must be port 1024 or above, unless changing jenkins user).
+    Type: Number
+    MinValue: 1024
+    MaxValue: 65535
+    Default: 8080
+  InstanceTagName:
+    Description: Instance Name tag that will be used to define the Name of the instance resource(s)
+    Type: String
+    Default: Jenkins
+  SnapshotId: 
+    Description: The Id of a snapshot (optional)
+    Default: ""
+    Type: String
+  VolumeSize: 
+    Description: Volum Size for Jenkins data volume that we will use for /var/lib/jenkins
+    Default: "25"
+    Type: String
+  PythonVersion: 
+    Description: Python Version
+    Default: "3.8.6"
+    Type: String
+  TerraformVersion: 
+    Description: Terraform Version
+    Default: "0.13.5"
+    Type: String
+
+# if SnapshotId is not empty, create volume from snapshot
+Conditions: 
+  UseSnapshot: !Not [!Equals [!Ref SnapshotId, ""]]
+
+# Create an easy mapping, simply mapping the region selected to the appropriate Amazon Linux 2 AMI
+Mappings: 
+  RegionMap: 
+    'ap-southeast-2': 
+      AMI: 'ami-076e39b6b14e3bb20'
```

Comment:
It is better to make AMI as a parameter.
Think from ops view, if Ops team will patch Jenkins OS using the latest image using your CF template, how to make their life easier?
In the first step, the ops team tears down the existing CF stack
When CF stack is deleted,  AWS  will create a snapshot for the data disk 
And Ops team create a new stack for the new AMI id and the snapshot id to build a new Jenkins box.

And can we improve the about procedure?

Reply:
Yeah, sounds reasonable.